src/lib/operators/insert.cpp

```diff
@@ -107,136 +93,132 @@ std::shared_ptr<const Table> Insert::_on_execute(std::shared_ptr<TransactionCont
+   *    Do so while locking the table to prevent multiple threads modifying the table's size simultaneously.
+   *    Since allocation is expected to be faster than writing to the memory, allocating under lock and then writing -
+   *    in a second step - without lock will minimize the time that the Table's append_mutex is locked.
+   */
   {
-    auto scoped_lock = _target_table->acquire_append_mutex();
+    const auto append_lock = _target_table->acquire_append_mutex();
+
+    auto remaining_rows = input_table_left()->row_count();
 
     if (_target_table->chunk_count() == 0) {
       _target_table->append_mutable_chunk();
     }
 
-    start_chunk_id = _target_table->chunk_count() - 1;
-    end_chunk_id = start_chunk_id + 1;
-    auto last_chunk = _target_table->get_chunk(start_chunk_id);
-    start_index = last_chunk->size();
+    while (remaining_rows > 0) {
+      auto target_chunk_id = ChunkID{_target_table->chunk_count() - 1};
+      auto target_chunk = _target_table->get_chunk(target_chunk_id);
 
-    // If last chunk is compressed, add a new uncompressed chunk
-    if (!last_chunk->is_mutable()) {
-      _target_table->append_mutable_chunk();
-      end_chunk_id++;
-    }
+      // If the last Chunk of the target Table is either immutable or full, append a new mutable Chunk
+      if (!target_chunk->is_mutable() || target_chunk->size() == _target_table->max_chunk_size()) {
+        _target_table->append_mutable_chunk();
+        ++target_chunk_id;
+        target_chunk = _target_table->get_chunk(target_chunk_id);
+      }
 
-    auto remaining_rows = total_rows_to_insert;
-    while (remaining_rows > 0) {
-      auto current_chunk = _target_table->get_chunk(static_cast<ChunkID>(_target_table->chunk_count() - 1));
-      auto rows_to_insert_this_loop = std::min(_target_table->max_chunk_size() - current_chunk->size(), remaining_rows);
+      const auto target_chunk_num_inserted_rows =
+          std::min<size_t>(_target_table->max_chunk_size() - target_chunk->size(), remaining_rows);
 
-      // Resize MVCC vectors.
-      current_chunk->get_scoped_mvcc_data_lock()->grow_by(rows_to_insert_this_loop, MvccData::MAX_COMMIT_ID);
+      _target_chunk_ranges.emplace_back(
+          ChunkRange{target_chunk_id, target_chunk->size(),
+                     static_cast<ChunkOffset>(target_chunk->size() + target_chunk_num_inserted_rows)});
 
-      // Resize current chunk to full size.
-      auto old_size = current_chunk->size();
-      for (ColumnID column_id{0}; column_id < current_chunk->column_count(); ++column_id) {
-        typed_segment_processors[column_id]->resize_vector(current_chunk->get_segment(column_id),
-                                                           old_size + rows_to_insert_this_loop);
-      }
+      // Grow MVCC columns while simultaneously locking them for the current transaction.
+      target_chunk->get_scoped_mvcc_data_lock()->grow_by(target_chunk_num_inserted_rows, context->transaction_id());
```

Comment:
Explain why we grow MVCC vectors before we touch the segments?

Reply:
The table is locked, no one else can resize its segments or mvcc-data.
Every query will see the data through the lenses of mvcc, it does not matter whether the mvcc data is not grown or whether it is grown but locked.

If anything I'd say we should resize the MVCC vectors AFTER the value segments, as the resize of the tid vector will be a release operations and "those should always happen at the end", informally speaking.

Reply:
@mrks I think we should take 5 minutes tomorrow to synchronize our understanding of the c++ memory model.

Reply:
Mmh, I thought this was done on purpose; so this was a request to add a comment.

>  Every query will see the data through the lenses of mvcc, it does not matter whether the mvcc data is not grown or whether it is grown but locked.

But how do the "lenses of MVCC" work? The validate operator looks at all rows (i.e., `chunks->size()`). If we resize the segments first, it might end up accessing non-existing MVCC info.

> I think we should take 5 minutes tomorrow to synchronize our understanding of the c++ memory model.

The C++ memory model in five minutes. I giggled.

Reply:
> Mmh, I thought this was done on purpose

c/p from the previous impl


> But how do the "lenses of MVCC" work? The validate operator looks at all rows (i.e., chunks->size()). If we resize the segments first, it might end up accessing non-existing MVCC info.

Then Validate *should* probably look at the mvcc-data-length... Really looking forward to tpcc with 100 clients to stress test all this...

Reply:
I think `Chunk::size` should return how many valid rows there are and no one should be allowed to touch any offsets in any vectors beyond that.

> Then Validate should probably look at the mvcc-data-length...

Which one of the three vectors?

Reply:
> Which one of the three vectors?

the tid vector, I guess. it should always be written to as the last operation and read from as the first operation - it contains atomics so this should get us closer to a safe acquire/release pattern

Reply:
(I am being vague here as really want to discuss this in the presence of a whiteboard...)

Reply:
Regarding 

> Explain why we grow MVCC vectors before we touch the segments? 

I added comments:

```c++
      // Grow MVCC vectors and mark new (but still empty) rows as being under modification by current transaction.
      // Do so before resizing the Segments, because the resize of `Chunk::_segments.front()` is what releases the
      // new row count.
      target_chunk->get_scoped_mvcc_data_lock()->grow_by(num_rows_for_target_chunk, context->transaction_id());

      // Grow data Segments.
      // Do so in REVERSE column order so that the resize of `Chunk::_segments.front()` happens last. It is this last
      // resize that makes the new row count visible to the outside world.
      auto old_size = target_chunk->size();
      for (ColumnID reverse_column_id{0}; reverse_column_id < target_chunk->column_count(); ++reverse_column_id) {
```